### PR TITLE
show the correct number of threads of agg in explain analyze

### LINKED ISSUE #5314: The number of threads of agg in explain analyze is not correct
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
TPC-H 10, explain analyze Q1

### 2. What did you expect to see? (Required)

The number of threads for agg is 20 according to the tiflash.log, the explain analyze result should also show that the number is 20.

### 3. What did you see instead (Required)
We see that the number of threads for agg is 1 in explain analyze result

`HashAgg_32`,`HashAgg_13`

```sql
> set tidb_isolation_read_engines = "tiflash_mpp,tiflash"; explain analyze /*PLACEHOLDER*/ select     l_returnflag,     l_linestatus,     sum(l_quantity) as sum_qty,     sum(l_extendedprice) as sum_base_price,     sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,     sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge,     avg(l_quantity) as avg_qty,     avg(l_extendedprice) as avg_price,     avg(l_discount) as avg_disc,     count(*) as count_order from     lineitem where     l_shipdate <= date_sub('1998-12-01', interval 108 day) group by     l_returnflag,     l_linestatus order by     l_returnflag,     l_linestatus;
+----------------------------------------+-------------+----------+--------------+----------------+--------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+---------+
| id                                     | estRows     | actRows  | task         | access object  | execution info                                                                             | operator info                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | memory  | disk    |
+----------------------------------------+-------------+----------+--------------+----------------+--------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+---------+
| Sort_6                                 | 2.98        | 4        | root         |                | time:3.26s, loops:2                                                                        | test.lineitem.l_returnflag, test.lineitem.l_linestatus                                                                                                                                                                                                                                                                                                                                                                                                                                                               | 11.5 KB | 0 Bytes |
| └─Projection_8                         | 2.98        | 4        | root         |                | time:3.26s, loops:2, Concurrency:OFF                                                       | test.lineitem.l_returnflag, test.lineitem.l_linestatus, Column#19, Column#20, Column#21, Column#22, Column#23, Column#24, Column#25, Column#26                                                                                                                                                                                                                                                                                                                                                                       | 11.4 KB | N/A     |
|   └─TableReader_36                     | 2.98        | 4        | root         |                | time:3.26s, loops:2, cop_task: {num: 1, max: 0s, proc_keys: 0, copr_cache_hit_ratio: 0.00} | data:ExchangeSender_35                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | N/A     | N/A     |
|     └─ExchangeSender_35                | 2.98        | 4        | mpp[tiflash] |                | tiflash_task:{time:3.25s, loops:1, threads:20}                                             | ExchangeType: PassThrough                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | N/A     | N/A     |
|       └─Projection_31                  | 2.98        | 4        | mpp[tiflash] |                | tiflash_task:{time:3.25s, loops:1, threads:20}                                             | Column#19, Column#20, Column#21, Column#22, div(Column#23, cast(case(eq(Column#43, 0), 1, Column#43), decimal(20,0) BINARY))->Column#23, div(Column#24, cast(case(eq(Column#44, 0), 1, Column#44), decimal(20,0) BINARY))->Column#24, div(Column#25, cast(case(eq(Column#45, 0), 1, Column#45), decimal(20,0) BINARY))->Column#25, Column#26, test.lineitem.l_returnflag, test.lineitem.l_linestatus                                                                                                                 | N/A     | N/A     |
|         └─HashAgg_32                   | 2.98        | 4        | mpp[tiflash] |                | tiflash_task:{time:3.25s, loops:1, threads:1}                                              | group by:test.lineitem.l_linestatus, test.lineitem.l_returnflag, funcs:sum(Column#46)->Column#19, funcs:sum(Column#47)->Column#20, funcs:sum(Column#48)->Column#21, funcs:sum(Column#49)->Column#22, funcs:sum(Column#50)->Column#43, funcs:sum(Column#51)->Column#23, funcs:sum(Column#52)->Column#44, funcs:sum(Column#53)->Column#24, funcs:sum(Column#54)->Column#45, funcs:sum(Column#55)->Column#25, funcs:sum(Column#56)->Column#26, funcs:firstrow(test.lineitem.l_returnflag)->test.lineitem.l_returnfla... | N/A     | N/A     |
|           └─ExchangeReceiver_34        | 2.98        | 4        | mpp[tiflash] |                | tiflash_task:{time:3.25s, loops:4, threads:20}                                             |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | N/A     | N/A     |
|             └─ExchangeSender_33        | 2.98        | 4        | mpp[tiflash] |                | tiflash_task:{time:3.25s, loops:256, threads:20}                                           | ExchangeType: HashPartition, Hash Cols: [name: test.lineitem.l_returnflag, collate: utf8mb4_bin], [name: test.lineitem.l_linestatus, collate: utf8mb4_bin]                                                                                                                                                                                                                                                                                                                                                           | N/A     | N/A     |
|               └─HashAgg_13             | 2.98        | 4        | mpp[tiflash] |                | tiflash_task:{time:3.25s, loops:256, threads:1}                                            | group by:Column#82, Column#83, funcs:sum(Column#72)->Column#46, funcs:sum(Column#73)->Column#47, funcs:sum(Column#74)->Column#48, funcs:sum(Column#75)->Column#49, funcs:count(Column#76)->Column#50, funcs:sum(Column#77)->Column#51, funcs:count(Column#78)->Column#52, funcs:sum(Column#79)->Column#53, funcs:count(Column#80)->Column#54, funcs:sum(Column#81)->Column#55, funcs:count(1)->Column#56                                                                                                             | N/A     | N/A     |
|                 └─Projection_41        | 59513965.06 | 58773499 | mpp[tiflash] |                | tiflash_task:{time:2.74s, loops:1047, threads:20}                                          | test.lineitem.l_quantity, test.lineitem.l_extendedprice, mul(test.lineitem.l_extendedprice, minus(1, test.lineitem.l_discount))->Column#74, mul(mul(test.lineitem.l_extendedprice, minus(1, test.lineitem.l_discount)), plus(1, test.lineitem.l_tax))->Column#75, test.lineitem.l_quantity, test.lineitem.l_quantity, test.lineitem.l_extendedprice, test.lineitem.l_extendedprice, test.lineitem.l_discount, test.lineitem.l_discount, test.lineitem.l_returnflag, test.lineitem.l_linestatus                       | N/A     | N/A     |
|                   └─Selection_30       | 59513965.06 | 58773499 | mpp[tiflash] |                | tiflash_task:{time:2.01s, loops:1047, threads:20}                                          | le(test.lineitem.l_shipdate, 1998-08-15 00:00:00.000000)                                                                                                                                                                                                                                                                                                                                                                                                                                                             | N/A     | N/A     |
|                     └─TableFullScan_29 | 60363016.00 | 58773499 | mpp[tiflash] | table:lineitem | tiflash_task:{time:2.01s, loops:1047, threads:20}                                          | keep order:false                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | N/A     | N/A     |
+----------------------------------------+-------------+----------+--------------+----------------+--------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+---------+
```

```
# tiflash.log
-- cop
[2022/07/07 19:50:51.997 +08:00] [DEBUG] [executeQuery.cpp:395] ["MPPTask:MPP<query:434423634838683652,task:1> Query pipeline:d
Union: <for mpp>d
 ExchangeSender x 20d
  Expression: <final projection>d
   Expression: <before order and select>d
    SharedQuery: <restore concurrency>d
     ParallelAggregating, max_threads: 20, final: trued
      Expression x 20: <before aggregation>d
       TiRemoteBlockInputStream(CoprocessorReader): schema: {<l_extendedprice, Decimal(15,2)>, <l_returnflag, String>, <l_linestatus, String>, <l_shipdate, MyDate>}d
       "] [thread_id=91]
```

### 4. What is your TiFlash version? (Required)

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #5314

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Show the correct number of threads of agg in explain analyze.
```

### CODE REVIEW

**Comment by SeaRise** [comment]
```
+----------------------------------+---------+---------+--------------+---------------+--------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------+--------+------+
| id                               | estRows | actRows | task         | access object | execution info                                                                             | operator info                                                                                                            | memory | disk |
+----------------------------------+---------+---------+--------------+---------------+--------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------+--------+------+
| TableReader_24                   | 12.00   | 5       | root         |               | time:1.44s, loops:2, cop_task: {num: 1, max: 0s, proc_keys: 0, copr_cache_hit_ratio: 0.00} | data:ExchangeSender_23                                                                                                   | N/A    | N/A  |
| └─ExchangeSender_23              | 12.00   | 5       | mpp[tiflash] |               | tiflash_task:{time:1.25s, loops:1, threads:20}                                             | ExchangeType: PassThrough                                                                                                | N/A    | N/A  |
|   └─Projection_22                | 12.00   | 5       | mpp[tiflash] |               | tiflash_task:{time:1.25s, loops:1, threads:20}                                             | Column#7                                                                                                                 | N/A    | N/A  |
|     └─HashAgg_8                  | 12.00   | 5       | mpp[tiflash] |               | tiflash_task:{time:1.24s, loops:1, threads:20}                                             | group by:test.test.col1, test.test.col3, funcs:count(1)->Column#7                                                        | N/A    | N/A  |
|       └─ExchangeReceiver_21      | 15.00   | 15      | mpp[tiflash] |               | tiflash_task:{time:1.24s, loops:1, threads:20}                                             |                                                                                                                          | N/A    | N/A  |
|         └─ExchangeSender_20      | 15.00   | 15      | mpp[tiflash] |               | tiflash_task:{time:1.24s, loops:1, threads:1}                                              | ExchangeType: HashPartition, Hash Cols: [name: test.test.col1, collate: binary], [name: test.test.col3, collate: binary] | N/A    | N/A  |
|           └─TableFullScan_19     | 15.00   | 15      | mpp[tiflash] | table:test    | tiflash_task:{time:1.24s, loops:1, threads:1}                                              | keep order:false, stats:pseudo                                                                                           | N/A    | N/A  |
+----------------------------------+---------+---------+--------------+---------------+--------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------+--------+------+
```

File: dbms/src/Flash/Statistics/ExecutorStatistics.h [hunk 1]
```diff
@@ -84,10 +84,11 @@ class ExecutorStatistics : public ExecutorStatisticsBase
         {
             for (const auto & input_stream : it->second)
             {
-                auto * p_stream = dynamic_cast<IProfilingBlockInputStream *>(input_stream.get());
-                assert(p_stream);
-                const auto & profile_info = p_stream->getProfileInfo();
-                base.append(profile_info);
+                if (auto * p_stream = dynamic_cast<IProfilingBlockInputStream *>(input_stream.get()))
```
Comment (JaySon-Huang) [inline dbms/src/Flash/Statistics/ExecutorStatistics.h:87]:
```suggestion
                if (auto * p_stream = dynamic_cast<IProfilingBlockInputStream *>(input_stream.get()); p_stream)
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 2]
```diff
@@ -451,6 +451,17 @@ void DAGQueryBlockInterpreter::recordProfileStreams(DAGPipeline & pipeline, cons
...
 
+void DAGQueryBlockInterpreter::recordProfileStream(const BlockInputStreamPtr & stream, size_t concurrency, const String & key)
+{
+    assert(stream);
+    auto & profile_streams = dagContext().getProfileStreamsMap()[key];
+    profile_streams.push_back(stream);
+    const auto & header = stream->getHeader();
+    /// We consider the number of streams as the concurrency of the operator and pass it to tidb.
+    for (size_t i = 1; i < concurrency; ++i)
+        profile_streams.emplace_back(std::make_shared<NullBlockInputStream>(header));
```
Comment (JaySon-Huang) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:462]: Do we fill the concurrency here?
https://github.com/pingcap/tiflash/blob/f97567ab67355768059fa93bc445e8ac5dbe5a39/dbms/src/Flash/Coprocessor/DAGResponseWriter.cpp#L94-L103

I don't get it that why using `NullBlockInputStream` here won't affect the number of actual rows, actual process time, etc.

Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:462]: because `NullBlockInputStream ` is not `IProfilingBlockInputStream`..

**Review by ywqzzy — APPROVED** [review]
LGTM

**Pull request closed without merging**